### PR TITLE
Tree-LSTM implementation for ChainerX

### PR DESCRIPTION
**Pull request by dido1998**
This depends on #7881

### CODE REVIEW

**Comment by dido1998** [comment]
Can this be reviewed?

**Comment by niboshi** [comment]
Some parts in this PR is overlapping with #7783. I'll merge this PR after that.
Triggering CIs for testing anyway.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Please resolve conflict.

**Comment by niboshi** [comment]
Thanks! Jenkins, test this please

There's a docs error in Travis CI. Please fix it.

**Comment by niboshi** [comment]
Additionally to the Travis error, Jenkins reports a clang-tidy error:
```
08:56:07 /repo/chainerx_cc/chainerx/routines/activation.cc:109:5: error: use range-based for loop instead [modernize-loop-convert,-warnings-as-errors]
08:56:07     for (size_t i = 0; i < fs.size(); i++) {
08:56:07     ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
08:56:07         (const auto & f : fs)
```

File: chainerx_cc/chainerx/routines/activation.cc [hunk 1]
```diff
@@ -91,6 +91,34 @@ Array LeakyRelu(const Array& x, Scalar slope) {
...
+    Array o = Squeeze(gates[2]);
+    std::vector<Array> fs;
+    for (size_t i = 3; i < gates.size(); i++) {
+        fs.emplace_back(Squeeze(gates[i]));
+    }
+    Array a_ = Tanh(a);
+    Array i_ = Sigmoid(i);
+    Array o_ = Sigmoid(o);
+    std::vector<Array> fs_s{};
+    for (const auto & f : fs) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/activation.cc]: I think this space should be removed (`auto &`)

Comment (dido1998) [reply chainerx_cc/chainerx/routines/activation.cc]: Done

**Comment by dido1998** [comment]
travis-ci is showing exceeded time-limit error

**Comment by niboshi** [comment]
Reran the test.
Jenkins test this please

**Comment by niboshi** [comment]
Jenkins test this please

**Pull request merged**